lib/bundles/inspec-compliance/target.rb

```diff
@@ -13,48 +13,80 @@ module Compliance
   class Fetcher < Fetchers::Url
     name 'compliance'
     priority 500
-    def self.resolve(target) # rubocop:disable PerceivedComplexity, Metrics/CyclomaticComplexity, Metrics/AbcSize
-      uri = if target.is_a?(String) && URI(target).scheme == 'compliance'
-              URI(target)
-            elsif target.respond_to?(:key?) && target.key?(:compliance)
-              URI("compliance://#{target[:compliance]}")
-            end
+    attr_reader :upstream_sha256
+
+    def initialize(target, opts)
+      super(target, opts)
+      if target.is_a?(Hash) && target.key?(:url)
+        @target = target[:url]
+        @upstream_sha256 = target[:sha256]
+      elsif target.is_a?(String)
+        @target = target
+        @upstream_sha256 = ''
+      end
+    end
+
+    def sha256
+      upstream_sha256.empty? ? super : upstream_sha256
+    end
+
+    def self.check_compliance_token(config)
+      if config['token'].nil? && config['refresh_token'].nil?
+        if config['server_type'] == 'automate'
+          server = 'automate'
+          msg = 'inspec compliance login https://your_automate_server --user USER --ent ENT --dctoken DCTOKEN or --token USERTOKEN'
+        elsif config['server_type'] == 'automate2'
+          server = 'automate2'
+          msg = 'inspec compliance login https://your_automate2_server --user USER --token APITOKEN'
+        else
+          server = 'compliance'
+          msg = "inspec compliance login https://your_compliance_server --user admin --insecure --token 'PASTE TOKEN HERE' "
+        end
+        raise Inspec::FetcherFailure, <<~EOF
+
+          Cannot fetch #{uri} because your #{server} token has not been
+          configured.
+
+          Please login using
+
+              #{msg}
+        EOF
+      end
+    end
 
+    def self.get_target_uri(target)
+      if target.is_a?(String) && URI(target).scheme == 'compliance'
+        URI(target)
+      elsif target.respond_to?(:key?) && target.key?(:compliance)
+        URI("compliance://#{target[:compliance]}")
+      end
+    end
+
+    def self.resolve(target)
+      uri = get_target_uri(target)
       return nil if uri.nil?
 
       # we have detailed information available in our lockfile, no need to ask the server
       if target.respond_to?(:key?) && target.key?(:url)
         profile_fetch_url = target[:url]
+        profile_checksum = target[:sha256]
```

Comment:
@chris-rock Regarding, backwards compatibility of lockfiles generated using the previous cache checksum method, as the comment here states, if the lockfile is present the profile will never be requested from the server (bypassing any possibility of using the new sha256 returned from the server). Instead, the checksum from the lockfile is passed, maintaining existing behavior for lockfiles. If you are using a version of inspec with these changes to re-vendor with --overwrite, the new sha256 will be used, and the old vendored dependency will be removed.

Reply:
This sounds reasonable. What about when you have a lock file but the cache has been removed? I think it may try to download and then compare but please confirm.

Reply:
With my latest commit, it will download the profile and use the checksum in the lockfile when the profile is not cached/vendored. I verified this works on A1 and A2.